awx/main/models/mixins.py

```diff
@@ -415,33 +415,41 @@ class TaskManagerJobMixin(TaskManagerUnifiedJobMixin):
     class Meta:
         abstract = True
 
-    def get_jobs_fail_chain(self):
-        return [self.project_update] if self.project_update else []
-
     def dependent_jobs_finished(self):
-        for j in self.dependent_jobs.all():
-            if j.status in ['pending', 'waiting', 'running']:
-                return False
-        return True
+        # if any dependent jobs are pending, waiting, or running, return False
+        return not any(j.status in ACTIVE_STATES for j in self.dependent_jobs.all())
 
 
 class TaskManagerUpdateOnLaunchMixin(TaskManagerUnifiedJobMixin):
     class Meta:
         abstract = True
 
-    def get_jobs_fail_chain(self):
-        return list(self.dependent_jobs.all())
-
 
 class TaskManagerProjectUpdateMixin(TaskManagerUpdateOnLaunchMixin):
     class Meta:
         abstract = True
 
+    def get_jobs_fail_chain(self):
+        return list(self.unifiedjob_blocked_jobs.all())
+
 
 class TaskManagerInventoryUpdateMixin(TaskManagerUpdateOnLaunchMixin):
     class Meta:
         abstract = True
 
+    def get_jobs_fail_chain(self):
+        blocked_jobs = list(self.unifiedjob_blocked_jobs.all())
+        other_updates = []
+        if blocked_jobs:
+            for dep in blocked_jobs[0].dependent_jobs.all():
```

Comment:
I'm a little bit suspicious of this. I don't know that the `blocked_jobs` list is ordered... it might be, but that's really non-obvious looking at it, and the 0 reference, `blocked_jobs[0]` always feels suspicious. If this is well-established, then that's fine, but some code comments could help. What _is_ the first entry of the blocked jobs?

If the list is always length 1, then maybe you should just loop instead of using the first one?

Reply:
oh that's probably the job. This is making more sense to me. It will fail its job if it fails... and any other unified jobs that is waiting on.

Reply:
yeah `blocked_jobs` are all the jobs that depend on this inventory update succeeding.

But, how can we fail the _other_ inventory updates that were started too (to retain the same behavior as devel)? Well, we can get that by looking at one of the blocked_jobs and getting its dependent_jobs. These dependent jobs might include project updates, so we need to filter it down further by checking its type

`if type(dep) is type(self) and dep.id != self.id:`

We can really grab _any_ of the blocked_jobs for this, as they all are guaranteed to have the same inventory, and thus the same set of inventory sources.

Reply:
add some comments around this @AlanCoding

Reply:
sounds good.

Ultimately, looks like these feed into `handle_work_error` via the dispatcher errback. That part of the design isn't bad, it's pretty sensible to handle the triggers this way. On the other hand, `handle_work_error` looks even worse than the reaper in terms for changing job status from underneath a running process, and not even bothering to send a cancel signal. Problem for another day...